### PR TITLE
Improve Sender Input Validation Tests and Error Types

### PR DESCRIPTION
**Pull request by Jolah1**
Part of #678.

## What

Check that every original sender outpoint is still present in the proposal, in its original order, before classifying any proposal input as the receiver's. Existing receiver-input tests now modify the proposal's actual receiver input instead of the sender's.

## Why

Master already rejects proposals with missing or shuffled sender inputs, but only after the input loop. An altered sender outpoint is classified as a receiver input first, so the sender sees an unrelated error ("an input in proposed transaction belonging to the receiver is not finalized"). Checking ordering up front makes the error name the actual problem.

The UTXO-data check was dropped per review; that part of #678 remains open.

## Tests

```sh
cargo test -p payjoin ordered_subsequence --all-features
cargo test -p payjoin test_sender_input_outpoint_changed --all-features
cargo test -p payjoin test_receiver_input_ --all-features
cargo test -p payjoin --all-features
cargo fmt --all -- --check
cargo clippy -p payjoin --all-targets --all-features -- -D warnings
codespell
```

Disclosure: code and PR text drafted with AI assistance (Codex, Claude); reviewed and tested by me.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MRG6vHFsY9Ny88eRtTQ2vz

### CODE REVIEW

**Review by caarloshenriq — COMMENTED** [review]
Concept ACK d576e05

The security logic is sound. The outpoint pre-check correctly uses a single advancing iterator with any() to enforce that all original outpoints appear as an ordered subsequence in the proposal, catching modified outpoints early instead of misclassifying them as receiver inputs. The UTXO validation with is_none_or correctly accepts stripped fields (per BIP 78) while rejecting modified or injected data. The fix to the existing receiver-input tests (operating on input[1] instead of mangling input[0]) is a genuine improvement: those tests now exercise the actual receiver input.

Two commit hygiene points:

This commit does three things: outpoint pre-check, UTXO data validation, and existing test fixes. Per project conventions each logical unit should be its own commit. At minimum the test fixes should be separate, since they're independently correct and useful regardless of the new validation.

The commit message has no body. For a security-relevant change to PSBT validation, it would be worth explaining the threat model (receiver modifying UTXO data to invalidate signatures or cause the sender to sign a different transaction). The context in #678 covers this well and the body could reference it.

**Comment by Jolah1** [comment]
> Concept ACK [d576e05](https://github.com/payjoin/rust-payjoin/commit/d576e051f1705cb91e9025f8bacd2835d4302370)
> 
> The security logic is sound. The outpoint pre-check correctly uses a single advancing iterator with any() to enforce that all original outpoints appear as an ordered subsequence in the proposal, catching modified outpoints early instead of misclassifying them as receiver inputs. The UTXO validation with is_none_or correctly accepts stripped fields (per BIP 78) while rejecting modified or injected data. The fix to the existing receiver-input tests (operating on input[1] instead of mangling input[0]) is a genuine improvement: those tests now exercise the actual receiver input.
> 
> Two commit hygiene points:
> 
> This commit does three things: outpoint pre-check, UTXO data validation, and existing test fixes. Per project conventions each logical unit should be its own commit. At minimum the test fixes should be separate, since they're independently correct and useful regardless of the new validation.
> 
> The commit message has no body. For a security-relevant change to PSBT validation, it would be worth explaining the threat model (receiver modifying UTXO data to invalidate signatures or cause the sender to sign a different transaction). The context in #678 covers this well and the body could reference it.

Thanks — split into three commits: the test fixes first (they pass on master's validation logic unchanged, so they stand alone as you said), then the outpoint pre-check, then the UTXO data validation with its error variant. 
  Each builds and passes cargo test --all-features + clippy independently; verified the three together reproduce the original tree exactly. Commit bodies now cover the threat model for each change and reference #678. Also 
  rebased onto current master while I was rewriting.

**Review by caarloshenriq — APPROVED** [review]
tACK bd2186a905eb6766e045be2b1fda476eb6956b7a

 All sender-input validation tests (test_sender_input_outpoint_changed, test_sender_input_witness_utxo_changed, test_sender_input_non_witness_utxo_changed) and the corrected receiver-input tests pass.

File: payjoin/src/core/send/mod.rs [hunk 1]
```diff
@@ -367,6 +367,14 @@ impl PsbtContext {
...
         ensure_receiver_input_finalized: bool,
     ) -> InternalResult<()> {
+        let mut proposed_inputs = proposal.unsigned_tx.input.iter();
+        ensure(
+            self.original_psbt.unsigned_tx.input.iter().all(|original| {
+                proposed_inputs.any(|proposed| proposed.previous_output == original.previous_output)
+            }),
+            InternalProposalError::MissingOrShuffledInputs,
+        )?;
+
```
Comment (xstoicunicornx) [inline payjoin/src/core/send/mod.rs]: Might be good to also have test for ordering? Rather than having to create test example with multiple sender inputs we could encapsulate this check in a function and test the function in a unit test. This would also be more readable.

For example function like: 
```rust
/// Whether `originals` appear among `proposed` in the same relative order.
///
/// BIP 78 forbids the receiver from shuffling: its additional inputs "must be
/// inserted at a random index", so the sender's inputs must survive as an
/// ordered subsequence of the proposal's.
fn is_ordered_subsequence<'a>(
    originals: impl IntoIterator<Item = &'a TxIn>,
    proposed: impl IntoIterator<Item = &'a TxIn>,
) -> bool {
    let mut proposed = proposed.into_iter();
    originals
        .into_iter()
        .all(|original| proposed.any(|p| p.previous_output == original.previous_output))
}
```

And this call site would be 

```suggestion
        ensure(
            is_ordered_subsequence(&self.original_psbt.unsigned_tx.input, &proposal.unsigned_tx.input),
            InternalProposalError::MissingOrShuffledInputs,
        )?;
```

Test for ordering can then be minimal and robust at same time:
```rust
#[test]
fn ordered_subsequence_requires_original_order() {
    let txin = |vout| TxIn {
        previous_output: OutPoint::new(Txid::all_zeros(), vout),
        ..Default::default()
    };
    let originals = vec![txin(0), txin(1)];

    assert!(is_ordered_subsequence(&originals, &vec![txin(0), txin(9), txin(1)]));
    assert!(!is_ordered_subsequence(&originals, &vec![txin(1), txin(0)]));
    assert!(!is_ordered_subsequence(&originals, &vec![txin(0)]));
}
```

File: payjoin/src/core/send/mod.rs [hunk 2]
```diff
@@ -404,6 +412,17 @@ impl PsbtContext {
...
+                        proposed
+                            .psbtin
+                            .witness_utxo
+                            .as_ref()
+                            .is_none_or(|utxo| Some(utxo) == original.psbtin.witness_utxo.as_ref())
+                            && proposed.psbtin.non_witness_utxo.as_ref().is_none_or(|utxo| {
+                                Some(utxo) == original.psbtin.non_witness_utxo.as_ref()
+                            }),
+                        InternalProposalError::SenderTxinUtxoInfoChanged,
+                    )?;
```
Comment (xstoicunicornx) [inline payjoin/src/core/send/mod.rs:449]: Can this be broken up into 2 separate `ensure`s?  Would make this more readable.

Also maybe worth having simple test for when utxo data is unchanged or none (can prob be combined into one test, test unchanged, wipe utxo data, test none).

**Review by xstoicunicornx — DISMISSED** [review]
utACK bd2186a905eb6766e045be2b1fda476eb6956b7a.

PR addresses some deficiencies in accurate error reporting when the sender is validating the payjoin proposal. None of these changes are absolutely necessary but do improve the clarity of our tests and error reporting.

Provided some small feedback.

File: payjoin/src/core/send/mod.rs (hunk 1 above)
Comment (Jolah1) [reply payjoin/src/core/send/mod.rs]: Done, extracted as you wrote it, with your test plus an in-order case. Kept it a free function so the ordering rule is testable without a proposal vector carrying several sender inputs.
 
   Folded into the outpoint commit rather than added on top, so each commit still stands alone.

File: payjoin/src/core/send/mod.rs (hunk 2 above)
Comment (Jolah1) [reply payjoin/src/core/send/mod.rs:449]: Split into two ensures, and added test_sender_input_utxo_info_unchanged_or_stripped covering both accepting cases in one test: unchanged data passes, then the fields are wiped and it passes again.
 
  Checked it isn't vacuous — swapping is_none_or for is_some_and makes it fail.

**Comment by Jolah1** [comment]
> utACK [bd2186a](https://github.com/payjoin/rust-payjoin/commit/bd2186a905eb6766e045be2b1fda476eb6956b7a).
> 
> PR addresses some deficiencies in accurate error reporting when the sender is validating the payjoin proposal. None of these changes are absolutely necessary but do improve the clarity of our tests and error reporting.
> 
> Provided some small feedback.

 updated bd2186a → 6c62fb2b: extracted is_ordered_subsequence with a unit test, split the UTXO check into two ensures, and added a test for unchanged/stripped data. Both folded into their existing commits, so the three-commit structure is unchanged. Re-ACK when you have a moment.

File: payjoin/src/core/send/mod.rs [hunk 3]
```diff
@@ -404,6 +427,21 @@ impl PsbtContext {
                             InternalProposalError::SenderTxinNonAllSighashType,
                         )?;
                     }
+                    // BIP 78 lets the receiver strip these fields, but never alter them.
```
Comment (xstoicunicornx) [inline payjoin/src/core/send/mod.rs]: BIP 78 doesn't currently say anything about these fields, we are enforcing this arbitrarily. Which I think is fine, but thinking about it more would like others to weigh in too.

File: payjoin/src/core/send/mod.rs [hunk 4]
```diff
@@ -747,6 +786,21 @@ mod test {
...
+    fn ordered_subsequence_requires_original_order() {
+        let txin = |vout| TxIn {
+            previous_output: OutPoint::new(Txid::all_zeros(), vout),
+            ..Default::default()
+        };
+        let originals = vec![txin(0), txin(1)];
+
+        assert!(is_ordered_subsequence(&originals, &vec![txin(0), txin(1)]));
+        // A receiver input inserted between the sender's is allowed.
+        assert!(is_ordered_subsequence(&originals, &vec![txin(0), txin(9), txin(1)]));
```
Comment (xstoicunicornx) [inline payjoin/src/core/send/mod.rs:804]:
```suggestion
        assert!(is_ordered_subsequence(&originals, &vec![txin(3), txin(0), txin(9), txin(1), txin(3)]));
        assert!(!is_ordered_subsequence(&originals, &vec![txin(3), txin(1), txin(9), txin(0), txin(3)]));
```

File: payjoin/src/core/send/mod.rs [hunk 5]
```diff
@@ -1241,6 +1295,75 @@ mod test {
...
+                InternalProposalError::SenderTxinUtxoInfoChanged.to_string()
+            );
+            Ok(())
+        }
+
+        #[test]
+        fn test_sender_input_utxo_info_unchanged_or_stripped() -> Result<(), BoxError> {
+            let mut ctx = create_psbt_context()?;
+            let previous_tx = PARSED_ORIGINAL_PSBT.clone().extract_tx_unchecked_fee_rate();
+            ctx.original_psbt.inputs[0].non_witness_utxo = Some(previous_tx.clone());
```
Comment (xstoicunicornx) [inline payjoin/src/core/send/mod.rs:1366]: What happened to `witness_utxo`?

File: payjoin/src/core/send/mod.rs [hunk 6]
```diff
@@ -1241,6 +1295,75 @@ mod test {
...
+            let mut proposal: bitcoin::Psbt = PARSED_PAYJOIN_PROPOSAL.clone();
+            proposal.inputs[0].non_witness_utxo = Some(previous_tx);
+
+            assert!(ctx.clone().process_proposal(proposal.clone()).is_ok());
+
+            // BIP 78 lets the receiver strip UTXO data it does not need.
```
Comment (xstoicunicornx) [inline payjoin/src/core/send/mod.rs]: Again, BIP 78 says nothing on this. Where are you getting this? Have you read BIP 78? Am I missing something?

File: payjoin/src/core/send/mod.rs (hunk 3 above)
Comment (Jolah1) [reply payjoin/src/core/send/mod.rs]: Agreed, this is policy, not spec, and the comment now says so. It also doesn't close a hole: restore_original_utxos overwrites these fields two lines after the check, so the sender signs its own values either way. The gain is a named error instead of a silent overwrite. Happy to drop this commit if the preference is status quo.

File: payjoin/src/core/send/mod.rs (hunk 4 above)
Comment (Jolah1) [reply payjoin/src/core/send/mod.rs:804]: Taken as written.

File: payjoin/src/core/send/mod.rs (hunk 5 above)
Comment (Jolah1) [reply payjoin/src/core/send/mod.rs:1366]: Missed it,  only non_witness_utxo was set. The vector's proposal happens to carry a matching witness_utxo, so the check wasn't vacuous, but that was luck. Asserted explicitly now.

File: payjoin/src/core/send/mod.rs (hunk 6 above)
Comment (Jolah1) [reply payjoin/src/core/send/mod.rs]: You're right, I was wrong — BIP 78 has no such rule, comment removed.
 
  Where I got it: restore_original_utxos on master says "The BIP78 spec requires utxo information to be removed," and I took it at face value. I can't find support for that either  no stripping language in the spec, 
 CreateOriginalPSBT clears only keypaths, partial sigs and unknowns, and both proposal vectors return the sender's witness_utxo unchanged.

File: payjoin/src/core/send/mod.rs (hunk 3 above)
Comment (xstoicunicornx) [reply payjoin/src/core/send/mod.rs]: Hmm... yes could we please drop this commit for now? Feel free to open as an issue to follow up on this question of whether we should enforce this policy even if not in the specification. Careful, I think some updates for the [Reject proposals missing original sender inputs](https://github.com/payjoin/rust-payjoin/pull/1836/changes/df46ca8218bbfece64a403c1f37acca983ef84e9) commit got entangled in this commit.

Comment (Jolah1) [reply payjoin/src/core/send/mod.rs]: > Hmm... yes could we please drop this commit for now? Feel free to open as an issue to follow up on this question of whether we should enforce this policy even if not in the specification. Careful, I think some updates for the [Reject proposals missing original sender inputs](https://github.com/payjoin/rust-payjoin/pull/1836/changes/df46ca8218bbfece64a403c1f37acca983ef84e9) commit got entangled in this commit.

Dropped, and moved the is_ordered_subsequence test cases back into the outpoint commit.

**Review by xstoicunicornx — APPROVED** [review]
utACK 283663e49b11a13030e284477dce628356b8eaa5

Improves a few tests by removing hacks that altered sender inputs just to get them to be treated as receiver inputs and instead just references the actual receiver inputs instead. Also improves accuracy in the error type returned when there are missing or shuffled sender inputs.

**Comment by Christopherdominic** [comment]
tACK 283663e

I traced the new `check_inputs()` flow and ran:

`cargo test -p payjoin --lib send::test`

All 40 tests passed, including the new `ordered_subsequence_requires_original_order` and `test_sender_input_outpoint_changed` tests.

The new check makes sense to me: it ensures the sender's original inputs are still present and ordered before the proposal inputs are classified, so altered sender outpoints get the appropriate `MissingOrShuffledInputs` error instead of being handled as receiver inputs.
